Add a tab showing the production summary

veger opened the pull request:
_It took some preparation and bug fix PRs, but this is (finally) the PR adding the 'Summary Tab'_

I tried to reuse the existing UI of YAFC, so I based the whole thing on a `ProjectPageContents` and a `ProjectPageView`, so it would be recognized and fit into the `MainScreen` (tabs) and serialization.

![YAFC-production-summary](https://user-images.githubusercontent.com/171827/147891150-b04e10a6-6088-4759-8299-f13449a5350f.png)


The UI of the new 'Summary' tab is ugly, I tried making it nicer, but I found it hard to understand the `ImGui` and its features and how to use them... So I propose this could be further improved om in separate PRs to get the basic functionality/feature into YAFC already and make it available for everyone (I used it a lot and cannot live without this anymore).

I also fixed some issues I found while implementing the Summary tab, moved to #31 and #34, except for:
* calculating `DatGrid` width when header is not used (I found the header useless here)

have-fun-was-taken commented:
For future bugfixing: The tab name overlaps with other information.
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/a345146c-10e5-4993-bcab-8a037b326be6)

have-fun-was-taken commented:
Do I understand it correctly that for now the user cannot scroll to the right?
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/88171b48-3cd6-4999-8df2-0d8818065119)

have-fun-was-taken commented:
When comparing your Summary that can be opened from the menu and the Production Summary that can be opened from the "+" sign, what is the difference between their purposes?

veger commented:
> For future bugfixing: The tab name overlaps with other information.

Hm... your screenshot looks different than what I am used to (except the dark vs light themes I guess).

![image](https://github.com/have-fun-was-taken/yafc-ce/assets/171827/cb95ff2e-fa6e-4944-9f1d-35311deb6d1e)



> Do I understand it correctly that for now the user cannot scroll to the right?

It was mentioned in #31 (IIRC): use CRL+scrollwheel or arrow key to scroll horizontally. For some reasons it is not possible to click on the scrollbar (something we might want to fix in the future).

> When comparing your Summary that can be opened from the top-left menu and the Production Summary that can be opened from the "+" sign, what is the difference between their purposes?

This is something that @ShadowTheAge started, but never finished.:
* https://github.com/ShadowTheAge/yafc/commit/155e7e6de3f6cec0710f33bd8c37cca44ccbe43c
* https://github.com/ShadowTheAge/yafc/commit/5357add72ded3d72dd8f8b482b3661e836850292
* https://github.com/ShadowTheAge/yafc/commit/75c7d98955e7b16648325ddf9931476b89ebdc43
* https://github.com/ShadowTheAge/yafc/commit/d39bbea677c2cd51fe8e3b53c08293898f02523f

Unfortunately, it is not in a feature branch, so finding all commits related to this is quite a bit of work.

If/when this PR is merged, we can start a cleanup of this WIP implementation.

have-fun-was-taken commented:
> [did this happen] when you make the Window too small for the tab bar?

I haven't understood what exactly you meant by "making it too small", but I didn't make it small on purpose. The app window was around 1500x900 at the time of making a screenshot. I attached the full-screen version below.

> I have short tab names

I think the tab names that have the length of something like "Military Science" should be handled without overlapping because that's a common name length.

![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/60d33f8b-aaca-4be6-9fe1-b63656c56672)

veger commented:
Oh now I see, it is the name on the left side of the Summary view, I thought it was on the tab bar that was wrapping.

IIRC, the width could not be made dynamic (or I was not able to do so, but I'll verify when I have some time)

have-fun-was-taken commented:
Also, I got a NullReferenceException when trying to delete the Summary page
Steps to reproduce:
1. Launch YAFC, make it full screen. 
2. Have a couple of sheets open. In my case, they were already opened as I was loading the existing project.
3. Open the Summary tab. Click "Only show issues", with ending on it being enabled.
4. Delete the Summary tab by clicking on it and choosing the delete option.
5. Get the error.
```
Object reference not set to an instance of an object.

   at YAFC.SummaryView.SummaryDataColumn.BuildElement(ImGui gui, ProjectPage page) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFC\Workspace\SummaryView.cs:line 76
   at YAFC.UI.DataGrid`1.BuildRow(ImGui gui, TData element, Single startX) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFC\Widgets\DataGrid.cs:line 168
   at YAFC.SummaryView.BuildScrollArea(ImGui gui) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFC\Workspace\SummaryView.cs:line 225
   at YAFC.UI.ImGui.DoGui(ImGuiAction action) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFCui\ImGui\ImGuiBuilding.cs:line 173
   at YAFC.UI.ImGui.MouseMove(Int32 mouseDownButton) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFCui\ImGui\ImGuiBuilding.cs:line 222
   at YAFC.UI.InputSystem.MouseMove(Int32 rawX, Int32 rawY) in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFCui\Core\InputSystem.cs:line 114
   at YAFC.UI.Ui.ProcessEvents() in C:\Users\Chill\Documents\git\have-fun\yafc-ce\YAFCui\Core\Ui.cs:line 112
```
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/247c5784-f596-4896-a8c2-5a962a601557)

veger commented:
>  I got a NullReferenceException when trying to delete the Summary page

I never tried this :rofl:

The Summary view is a singleton, as it is a special production page. So it should not be fully deleted (hence the error).
Instead, it gets hidden now, like the close button on the tab got clicked.

veger commented:
> the width could not be made dynamic (or I was not able to do so, but I'll verify when I have some time)

I checked and it is possible to set a min/max width for the column. But it does not work when I make the `maxWidth` larger than the (inital) width. After I quick code inspection, I do see some code that is supposed to calculate the width, but it does not seem to be working...?

I have a plan to move the first column outside of the DataGrid, but I got some headaches as it still would need to scroll vertically! (we'd need to two/nested ScrollAreas for that...)

Either way, I do feel there is a bug in the DataGrid and calculating dynamic widths. Most stuff in YAFC is using static widths, so I guess this bug is not triggered until now when I tried to set the min/max allowed width.

have-fun-was-taken commented:
I think this feature will take quite some time to polish and then replace the current Summary with. So I think about releasing YAFC with the current features, and moving this one to the next release. @veger what do you think?

veger commented:
It is fine by me, we can focus for the release after this include this overview/summary tab.

have-fun-was-taken commented:
Alright! Then I'll playtest the current version for a bit and then release.

veger commented:
I rebased and reformatted (for the new editorconfig) this PR.

## Known issues (summary):

> NullReferenceException when trying to delete the Summary page

Fixed

> The tab name overlaps with other information

The data grid does not support dynamic widths (although the code looks like it should).
I think the grid needs to be adjusted to have heading (top and left) support. Then it is able to calculate the required width and adjust accordingly.
Or we could even add such a support for each row/column to have automatic widths/heights for the whole grid (no idea if this is (too) computation intensive or something though)

Nonetheless, I guess we should open an issue for this (bug), and discuss the wished outcome/support of the data grid?

For this PR I could increase the width, but it won't help too much...

> Do I understand it correctly that for now the user cannot scroll to the right?

\<CRTL\>+\<scroll wheel\> or arrow keys do work.

Clicking the scrollbar does not work. Back then I tried to figure out what is causing this, but was not able to.
I suppose it is some YAFCui/SDL issue.

Let's open a ticket for this as well?

----

Is there anything else known that needs solving?

veger commented:
> Is there anything else known that needs solving?

Well... there are a lot of improvements imaginable, but I would propose to add new features for the Summay Tab in future PRs :wink:

have-fun-was-taken commented:
> The data grid does not support dynamic widths

What do you think about calculating the max length of the name column before generating the page, and setting it to that number? Is that a dynamic width you meant, or can it be arranged?

shihan42 commented:
> The data grid does not support dynamic widths (although the code looks like it should). I think the grid needs to be adjusted to have heading (top and left) support. Then it is able to calculate the required width and adjust accordingly. Or we could even add such a support for each row/column to have automatic widths/heights for the whole grid (no idea if this is (too) computation intensive or something though)
> 
> Nonetheless, I guess we should open an issue for this (bug), and discuss the wished outcome/support of the data grid?

Discussion is a good idea. Right know, I'm not sure how that could be solved or in what way the data grid should change.

> Clicking the scrollbar does not work. Back then I tried to figure out what is causing this, but was not able to. I suppose it is some YAFCui/SDL issue.
> Let's open a ticket for this as well?

Good idea, too. It is more than irritating that there are scrollbars but only one of them works.

veger commented:
> What do you think about calculating the max length of the name column before generating the page, and setting it to that number? Is that a dynamic width you meant, or can it be arranged?

It could be done like this (yes, it makes the width dynamic) but that is moving UI related stuff to the model/analyzer code.

Ideally, the DataGrid UI should do this automatically without the non-UI part (or at least the non-low level UI part) of YAFC to know about this calculations.

> Discussion is a good idea. Right know, I'm not sure how that could be solved or in what way the data grid should change.

Me neither, I feel there is a bug, or unfinished code in the DataGrid.

> Good idea, too. It is more than irritating that there are scrollbars but only one of them works.

I'll make a ticket, so other know how to use it. And for us to remember to figure this out and fix it

have-fun-was-taken commented:
Then I would suggest to make the length of the first column so that it can fit at least 20 characters.

veger commented:
> 20 characters.

Sure, I'll use a `o` for the width.

As the `w` and `i` have different (extreme) widths, 20 `w` would be too wide and 20 `i` too narrow, I suppose.

shihan42 reviewed (commented):
Did you take into account that these measurements are some logical measurements and are converted to real pixels at a late step during UI drawing?
Thats something I fell for, at first.

veger commented:
I used the low-tech approach: fill in some value and see whether it was wide enough.

After 5 ( :disappointed: ) iterations I could fit 20 `o` without too much empty space after it :wink:
